ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs

```diff
@@ -191,42 +190,34 @@ instance TPraosCrypto c => UpdateLedger (ShelleyBlock c) where
 instance TPraosCrypto c => LedgerSupportsProtocol (ShelleyBlock c) where
   protocolLedgerView _cfg = SL.currentLedgerView . shelleyState
 
-  anachronisticProtocolLedgerView_ cfg ledgerState slot =
-      case History.find slot history of
-        Just lv -> return lv
-        Nothing
-          | slot <  At maxLo
-          -> throwError TooFarBehind -- lower bound is inclusive
-          | slot >= At maxHi
-          -> throwError TooFarAhead  -- upper bound is exclusive
-          | otherwise
-          -> -- If 'futureLedgerView' fails, it is a bug
-             return $ either (error "futureLedgerView failed") id $
-             SL.futureLedgerView globals shelleyState forSlot
+  ledgerViewForecastAt_ cfg ledgerState at = do
+      guard (at >= minLo)
+      return $ Forecast at $ \for ->
+        case History.find (At for) history of
+          Just lv -> return lv
+          Nothing -> do
+            when (for >= maxHi) $
+              throwError $ OutsideForecastRange at for
+            -- If 'futureLedgerView' fails, it is a bug
```

Comment:
Document why bounds set by `futureLedgerView` are wrong (too far into the future: based on current ledger state, not based on intersection point).